main/src/mill/main/MainModule.scala

```diff
@@ -348,12 +348,25 @@ trait MainModule extends BaseModule0 {
               paths :+ potentialModulePath
             } else paths
           }
+          (allPaths, ts)
         }
 
     pathsToRemove match {
       case Left(err) =>
         Result.Failure(err)
-      case Right(paths) =>
+      case Right((paths, allSegments)) =>
+        val workersToRemove = evaluator.workerCache
+          .keysIterator
+          .filter(workerSegments => allSegments.exists(workerSegments.startsWith))
+          .toVector
+        val closeableCleanedWorkers = workersToRemove
+          .flatMap(workerSegments => evaluator.mutableWorkerCache.remove(workerSegments))
+          .collect {
+            case (_, Val(closeable: AutoCloseable)) =>
+              closeable
+          }
+        closeableCleanedWorkers.foreach(_.close())
```

Comment:
This is kind of roundabout, probably would be cleaner to write as a straight loop

```scala
for{
  workerSegments <- evaluator.workerCache.keysIterator.toList
  if allSegments.exists(workerSegments.startsWith))
}{
  evaluator.mutableWorkerCache.remove(workerSegments) match{
    case (_, Val(closeable: AutoCloseable)) => closeable.close()
    case _ => // do nothing
  }
}
```

Reply:
Sure